rclpy/src/rclpy/_rclpy.c

```diff
@@ -2601,6 +2604,48 @@ rclpy_wait(PyObject * Py_UNUSED(self), PyObject * args)
   Py_RETURN_NONE;
 }
 
+/// Take a raw message from a given subscription (internal- for rclpy_take with raw=True)
+/**
+ * \param[in] rcl subscription pointer pointing to the subscription to process the message
+ * \return Python byte array with the raw serialized message contents
+ */
+static PyObject *
+rclpy_take_raw(rcl_subscription_t * subscription)
+{
+  // Create a serialized message object
+  rcl_serialized_message_t msg = rmw_get_zero_initialized_serialized_message();
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rcl_ret_t ret = rmw_serialized_message_init(&msg, 0u, &allocator);
+  if (ret != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to initialize message: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
+    if (r_fini != RMW_RET_OK) {
+      PyErr_Format(PyExc_RuntimeError, "Failed to deallocate message buffer: %d", r_fini);
+    }
+    return NULL;
+  }
+
+  ret = rcl_take_serialized_message(subscription, &msg, NULL);
+  if (ret != RMW_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
+      "Failed to take_serialized from a subscription: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+    rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
+    if (r_fini != RMW_RET_OK) {
+      PyErr_Format(PyExc_RuntimeError, "Failed to deallocate message buffer: %d", r_fini);
+    }
+    return NULL;
+  }
+  PyObject * python_bytes = PyBytes_FromStringAndSize((char *)(msg.buffer), msg.buffer_length);
+  rmw_ret_t r_fini = rmw_serialized_message_fini(&msg);
+  if (r_fini != RMW_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError, "Failed to deallocate message buffer: %d", r_fini);
```

Comment:
should it return `NULL` in this case?

Reply:
It now returns NULL after decremented the `python_bytes` refcount in this error condition. If this is an adequate solution, I think this might finally be merge-able :)